include/net/p4tc.h

```diff
@@ -156,33 +156,58 @@ struct p4tc_table_key {
+
+#define P4TC_PERMISSIONS_CONTROL (GENMASK(9, 5))
+#define P4TC_PERMISSIONS_DATA (GENMASK(4, 0))
+
+struct p4tc_table_defact {
+	struct tc_action **default_acts;
+	/* Will have 2 5 bits blocks containing CRUDX (Create, read, update,
+	 * delete, execute) permissions for control plane and data plane.
+	 * The first 5 bits are for control and the next five are for data plane.
+	 * |crudxcrudx| if we were to denote it as UNIX permission flags.
+	 */
+	__u16 permissions;
+	struct rcu_head  rcu;
+};
+
 struct p4tc_table {
-	struct p4tc_template_common common;
-	struct idr                  tbl_keys_idr;
-	struct idr                  tbl_masks_idr;
-	struct idr                  tbl_prio_idr;
-	struct p4tc_cmd_value_ops   tbl_value_ops;
-	struct rhltable             tbl_entries;
-	struct tc_action            **tbl_preacts;
-	int                         tbl_num_preacts;
-	struct tc_action            **tbl_postacts;
-	struct tc_action            **tbl_default_hitact;
-	struct tc_action            **tbl_default_missact;
-	spinlock_t                  tbl_masks_idr_lock;
-	spinlock_t                  tbl_prio_idr_lock;
-	int                         tbl_num_postacts;
-	u32                         tbl_count;
-	u32                         tbl_curr_count;
-	u32                         tbl_keysz;
-	u32                         tbl_id;
-	u32                         tbl_keys_count;
-	u32                         tbl_max_entries;
-	u32                         tbl_max_masks;
-	u32                         tbl_curr_used_entries;
-	u32                         tbl_default_key;
-	refcount_t                  tbl_ctrl_ref;
-	refcount_t                  tbl_ref;
-	refcount_t                  tbl_entries_ref;
+	struct p4tc_template_common    common;
+	struct idr                     tbl_keys_idr;
+	struct idr                     tbl_masks_idr;
+	struct idr                     tbl_prio_idr;
+	struct p4tc_cmd_value_ops      tbl_value_ops;
+	struct rhltable                tbl_entries;
+	struct tc_action               **tbl_preacts;
+	int                            tbl_num_preacts;
+	struct tc_action               **tbl_postacts;
+	struct p4tc_table_defact __rcu *tbl_default_hitact;
+	struct p4tc_table_defact __rcu *tbl_default_missact;
+	spinlock_t                     tbl_masks_idr_lock;
+	spinlock_t                     tbl_prio_idr_lock;
+	int                            tbl_num_postacts;
+	u32                            tbl_count;
+	u32                            tbl_curr_count;
+	u32                            tbl_keysz;
+	u32                            tbl_id;
+	u32                            tbl_keys_count;
+	u32                            tbl_max_entries;
+	u32                            tbl_max_masks;
+	u32                            tbl_curr_used_entries;
+	u32                            tbl_default_key;
+	refcount_t                     tbl_ctrl_ref;
+	refcount_t                     tbl_ref;
+	refcount_t                     tbl_entries_ref;
 };
 
```

Comment:
does pahole like this struct?

Reply:
I changed the positions of two fields and this was the result:

```
struct p4tc_table {
        struct p4tc_template_common common;              /*     0   272 */
        /* --- cacheline 4 boundary (256 bytes) was 16 bytes ago --- */
        struct idr                 tbl_keys_idr;         /*   272    88 */
        /* --- cacheline 5 boundary (320 bytes) was 40 bytes ago --- */
        struct idr                 tbl_masks_idr;        /*   360    88 */
        /* --- cacheline 7 boundary (448 bytes) --- */
        struct idr                 tbl_prio_idr;         /*   448    88 */
        /* --- cacheline 8 boundary (512 bytes) was 24 bytes ago --- */
        struct p4tc_cmd_value_ops  tbl_value_ops;        /*   536     8 */
        struct rhltable            tbl_entries;          /*   544   352 */
        /* --- cacheline 14 boundary (896 bytes) --- */
        struct tc_action * *       tbl_preacts;          /*   896     8 */
        struct tc_action * *       tbl_postacts;         /*   904     8 */
        struct p4tc_table_defact * tbl_default_hitact;   /*   912     8 */
        struct p4tc_table_defact * tbl_default_missact;  /*   920     8 */
        spinlock_t                 tbl_masks_idr_lock;   /*   928    64 */
        /* --- cacheline 15 boundary (960 bytes) was 32 bytes ago --- */
        spinlock_t                 tbl_prio_idr_lock;    /*   992    64 */
        /* --- cacheline 16 boundary (1024 bytes) was 32 bytes ago --- */
        int                        tbl_num_postacts;     /*  1056     4 */
        int                        tbl_num_preacts;      /*  1060     4 */
        u32                        tbl_count;            /*  1064     4 */
        u32                        tbl_curr_count;       /*  1068     4 */
        u32                        tbl_keysz;            /*  1072     4 */
        u32                        tbl_id;               /*  1076     4 */
        u32                        tbl_keys_count;       /*  1080     4 */
        u32                        tbl_max_entries;      /*  1084     4 */
        /* --- cacheline 17 boundary (1088 bytes) --- */
        u32                        tbl_max_masks;        /*  1088     4 */
        u32                        tbl_curr_used_entries; /*  1092     4 */
        u32                        tbl_default_key;      /*  1096     4 */
        refcount_t                 tbl_ctrl_ref;         /*  1100     4 */
        refcount_t                 tbl_ref;              /*  1104     4 */
        refcount_t                 tbl_entries_ref;      /*  1108     4 */

        /* size: 1112, cachelines: 18, members: 26 */
        /* last cacheline: 24 bytes */
};
```

Reply:
Maybe premature but you should put the most frequently accessed fields in the datapath at the top.
Off top of my head tbl_pre/postacts, tbl_default_hit/missact.

Reply:
Opened an issue for it: https://github.com/tammela/linux-p4tc/issues/128